internal/controller/genericprovider_controller.go

```diff
@@ -127,7 +140,6 @@ func patchProvider(ctx context.Context, provider genericprovider.GenericProvider
 func (r *GenericProviderReconciler) reconcile(ctx context.Context, provider genericprovider.GenericProvider, genericProviderList genericprovider.GenericProviderList) (ctrl.Result, error) {
 	reconciler := newPhaseReconciler(*r, provider, genericProviderList)
 	phases := []reconcilePhaseFn{
-		reconciler.preflightChecks,
```

Comment:
What if we kept the preflightChecks as a phase here so we can benefit from the centralised condition setting logic, and rather we defined the `halt` pattern you came up with, as a pattern for all the phases.

Maybe with a customisable halting message (similar to the one you specified [here](https://github.dev/Fedosin/cluster-api-operator/blob/b17b9c91be17e774101047058762226f192bf96c/internal/controller/genericprovider_controller.go#L117-L118)) that can be logged before returning reconcile()? We could even have `halt` as a message string ()`haltMsg`?), and when the phase returns a non empty msg then we log and halt?

So [here](https://github.com/Fedosin/cluster-api-operator/blob/5c223b751ebc4c19fc4bdcc0ba8027ecbbd56fdb/internal/controller/genericprovider_controller.go) doing something like
```
res, halt, err = phase(ctx)
...
```